jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/EventsHandler.java

```diff
@@ -0,0 +1,419 @@
+     */
+    protected void onResponseWriteComplete(Request request, Throwable failure)
+    {
+    }
+
+    /**
+     * Invoked after the response trailers have been written <em>and</em> the final {@link #onResponseWriteComplete(Request, Throwable)} event was fired.
+     *
+     * @param request the request object. The {@code read()}, {@code demand(Runnable)} and {@code fail(Throwable)} methods must not be called by the listener.
+     * @param trailers the written trailers.
+     */
+    protected void onResponseTrailersComplete(Request request, HttpFields trailers)
+    {
+    }
+
+    /**
+     * Invoked when the request <em>and</em> response processing are complete,
+     * just before the request and response will be recycled (i.e. after the
+     * {@link Runnable} return from {@link org.eclipse.jetty.server.HttpChannel#onRequest(MetaData.Request)}
+     * has returned and the {@link Callback} passed to {@link Handler#handle(Request, Response, Callback)}
+     * has been completed).
+     *
+     * @param request the request object. The {@code read()}, {@code demand(Runnable)} and {@code fail(Throwable)} methods must not be called by the listener.
+     * @param failure if there was a failure to complete
+     */
+    protected void onComplete(Request request, Throwable failure)
+    {
+    }
+
+    private class EventsResponse extends Response.Wrapper
+    {
+        private boolean notifiedOnResponseBegin;
+        private HttpFields suppliedTrailers;
+
+        public EventsResponse(ReadOnlyRequest request, Response response)
+        {
+            super(request, response);
+        }
+
+        @Override
+        public void write(boolean last, ByteBuffer byteBuffer, Callback callback)
+        {
+            if (!notifiedOnResponseBegin)
+            {
+                fireOnResponseBegin(getRequest(), getWrapped().getHeaders().asImmutable(), getWrapped().getStatus());
+                notifiedOnResponseBegin = true;
+            }
+            fireOnResponseWrite(getRequest(), last, byteBuffer);
+            super.write(last, byteBuffer, Callback.from(callback, (x) -> fireOnResponseWriteComplete(getRequest(), x)));
+        }
+
+        @Override
+        public void setTrailersSupplier(Supplier<HttpFields> trailers)
+        {
+            super.setTrailersSupplier(trailers == null ? null : () -> suppliedTrailers = trailers.get());
+        }
+    }
+
+    private class EventsRequest extends Request.Wrapper
+    {
+        private final ReadOnlyRequest roRequest;
+
+        public EventsRequest(Request request, ReadOnlyRequest roRequest)
+        {
+            super(request);
+            this.roRequest = roRequest;
+        }
+
+        @Override
+        public Content.Chunk read()
+        {
+            Content.Chunk chunk = super.read();
+            fireOnRequestRead(roRequest, chunk);
+            return chunk;
+        }
+    }
+
+    private static class ReadOnlyRequest extends Request.Wrapper
```

Comment:
This should be moved to Request.ReadOnly as it plausible that it will be used elsewhere.

Reply:
I've created a `default Request asReadOnly()` method on `Request` instead. Good with you?

Reply:
I'm not sure we want another method on the Request that can be wrapped.   A static method would be OK!